### PR TITLE
Certik bridge / BLS findings. Batch 2 

### LINKED ISSUE #754: Certik, Ethereum Bridge, Preliminary Report (v1), Severity: Centralization [Priority 2]
- [ ] [GEB-54 | Centralization Risks in Smart Contracts](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-54)

### LINKED ISSUE #755: [P1] Certik, Ethereum Bridge, Preliminary Report (v1), Severity: Major [Priority 3]
- [x] GEB-03 | Duplicate Slot Indices Inflate Threshold Coverage - #822 
- [x] GEB-12 | Dealer Commitments Not Bound to Threshold Degree - #822 
- [x] GEB-29 | Weak Dealer Approval Enables Threshold Signing DoS] - #988 
- [x] GEB-34 | Genesis Epoch's Group Key Can Be Set by External Users - https://github.com/gonka-ai/gonka/pull/949

### LINKED ISSUE #756: Certik, Ethereum Bridge, Preliminary Report (v1), Severity: Medium [Priority 4]
- [ ] [GEB-04 | Incorrect Signing Threshold in `checkThresholdAndAggregate()`](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-04)
- [ ] [GEB-05 | Native Denom Auto-Detection Can Be Misconfigured in `community-sale` Contract](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-05)
- [ ] [GEB-13 | Aggregation of BLS Partial Signature Does Not Eliminate Duplicates](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-13)
- [ ] [GEB-14 | User-Controlled RequestId Allows Front-Run Poisoning of Threshold Signing](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-14)
- [ ] [GEB-15 | Cross-Chain Address Collision](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-15)
- [ ] [GEB-16 | Bridge BLS Signatures Are Not Bound to Destination Contract](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-16)
- [ ] [GEB-17 | Dealer Validation Majority Is Too Weak for Safe Key Recovery](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-17)
- [ ] [GEB-35 | Secret Shares Not Using Consensus `ValidDealers`](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-35)
- [ ] [GEB-36 | Authority Mismatch In `MigrateAllWrappedTokenContracts()`](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-36)
- [ ] [GEB-55 | BLS Genesis Export/Import Drops In-flight DKG and Signing state](https://skyharbor.certik.com/shared-report/c4dcdbdc-c10d-40c7-a8ff-894e13a1cb17?findingIndex=GEB-55)

### LINKED ISSUE #757: [P1] Certik, Ethereum Bridge, Preliminary Report (v1), Severity: Minor [Priority 5]
- [x] GEB-06 | Known Security Issue in Upstream Dependencies - https://github.com/gonka-ai/gonka/pull/675
- [x] GEB-07 | Weak Address Validation in `withdraw()` in `wrapped-token` Contract - https://github.com/gonka-ai/gonka/pull/814
- [x] GEB-08 | `ADMIN` Role Cannot Not Be Updated https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-09 | Migration From cw20‑base Leaves Required Wrapped‑Token State Uninitialized https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-10 | Migration of `community‑sale` Lacks Compatibility Checks and State Validation - https://github.com/gonka-ai/gonka/pull/814
- [x] GEB-18 | Slot Donation Picks Under-Allocated Donor, Enabling Sybil Weight Inflation - #822 
- [x] GEB-19 | Secret Shares Logged in `logging.Debug` - #822 
- [x] GEB-20 | Decoding Returns Zero If Fails https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-21 | Ineffective Polynomial Degree Check in `evaluatePolynomial()` https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-22 | Unchecked `amountToBytes32()` Panic on Oversized Amounts https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-23 | Missing Validation of `MsgRequestThresholdSignature.ValidateBasic()` for `chain_id`/`request_id` and Data Chunk Sizes - #822 
- [x] GEB-24 | Insufficient Validation for Dealer Part Submissions in `MsgSubmitDealerPart.ValidateBasic()` - #822 
- [x] GEB-25 | Missing Validation in Group Key Validation Signatures in `MsgSubmitGroupKeyValidationSignature.ValidateBasic()` - #822 
- [x] GEB-26 | Missing Validation in Partial Signature Submissions in `MsgSubmitPartialSignature.ValidateBasic()` - #822 
- [x] GEB-27 | Unbounded `DealerValidity` in Verification Vector Submissions of `MsgSubmitVerificationVector.ValidateBasic()` - #822 
- [x] GEB-37 | DKG Process Can Be Stuck Due to Internal Errors https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-38 | Inconsistent Comparison of Deadline Block https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-39 | Missing Validation of `msg.Amount` Being Positive in `MsgRequestBridgeWithdrawal` - https://github.com/gonka-ai/gonka/pull/814
- [x] GEB-40 | Broken Cleanup Logic https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-42 | Unhandled Error of `EmitTypedEvent()` https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-43 | Valid Dealers Can Be Less Than Threshold https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-47 | Hard-coded threshold for BLS signature https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-48 | Missing Signed Status in `parseEpochDataFromJSON()` https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-49 | Missing Check of Withdrawal and Mint Amount https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-56 | Missing Validation of Epoch Id in `RequestThresholdSignature()` https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-57 | Slot Range Silently Clamped Instead of Failing https://github.com/gonka-ai/gonka/pull/949

### LINKED ISSUE #758: [P1] Certik, Ethereum Bridge, Preliminary Report (v1), Severity: Informational [Priority 6]
- [x] GEB-11 | `InstantiateMsg.marketing` Is Ignored - https://github.com/gonka-ai/gonka/pull/814
- [x] GEB-50 | Unused Function `computeParticipantPublicKey()` in `bls_crypto.go` https://github.com/gonka-ai/gonka/pull/949
- [x] GEB-58 | `ProcessThresholdSigningRequested()` Incorrectly Returns Error https://github.com/gonka-ai/gonka/pull/949

### PR DESCRIPTION
**Pull request by GLiberman**
**Certik Ethereum Bridge**: Severities and titles are taken from the audit.

| ID | Severity | Title | Fixed in |
|:---|:---|:---|:---|
| GEB-55 | Medium | BLS Genesis Export/Import Drops In-flight DKG and Signing state | `inference-chain/proto/inference/bls/genesis.proto`<br>`inference-chain/x/bls/keeper/genesis.go`<br>`inference-chain/x/bls/module/genesis.go`<br>`inference-chain/x/bls/module/genesis_test.go`<br>`inference-chain/x/bls/types/genesis.go`<br>`inference-chain/x/bls/types/genesis.pb.go`<br>`inference-chain/x/bls/types/genesis_test.go`<br>`inference-chain/x/bls/types/keys.go`<br>`inference-chain/x/bls/keeper/msg_server_group_validation.go` |
| GEB-34 | Major | Genesis Epoch's Group Key Can Be Set by External Users | `proposals/ethereum-bridge-contact/contracts/BridgeContract.sol` |
| GEB-16 | Medium | Bridge BLS Signatures Are Not Bound to Destination Contract | `inference-chain/proto/inference/inference/tx.proto`<br>`inference-chain/api/inference/bls/tx.pulsar.go`<br>`inference-chain/api/inference/inference/tx.pulsar.go`<br>`inference-chain/x/inference/keeper/bridge_signature_internal_test.go`<br>`inference-chain/x/inference/keeper/epoch.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_mint.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_mint_test.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_withdrawal.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_withdrawal_test.go`<br>`inference-chain/x/inference/types/message_request_bridge_mint.go`<br>`inference-chain/x/inference/types/message_request_bridge_withdrawal.go`<br>`inference-chain/x/inference/types/tx.pb.go`<br>`proposals/ethereum-bridge-contact/contracts/BridgeContract.sol` |
| GEB-14 | Medium | User-Controlled RequestId Allows Front-Run Poisoning of Threshold Signing | `inference-chain/proto/inference/bls/tx.proto`<br>`inference-chain/x/bls/keeper/msg_server_threshold_signing.go`<br>`inference-chain/x/bls/keeper/msg_server_threshold_signing_test.go`<br>`inference-chain/x/bls/types/tx.pb.go` |
| GEB-57 | Minor | Slot Range Silently Clamped Instead of Failing | `inference-chain/x/bls/keeper/dkg_initiation.go` |
| GEB-56 | Minor | Missing Validation of Epoch Id in `RequestThresholdSignature()` | `inference-chain/testutil/keeper/expected_keepers_mocks.go`<br>`inference-chain/x/bls/keeper/keeper.go`<br>`inference-chain/x/bls/keeper/msg_server_threshold_signing.go`<br>`inference-chain/x/bls/keeper/msg_server_threshold_signing_test.go`<br>`inference-chain/x/inference/keeper/collateral_integration_test.go`<br>`inference-chain/x/inference/keeper/epoch.go`<br>`inference-chain/x/inference/keeper/epoch_test.go`<br>`inference-chain/x/inference/keeper/streamvesting_integration_test.go`<br>`inference-chain/x/inference/module/genesis_test.go`<br>`inference-chain/x/inference/types/expected_keepers.go` |
| GEB-47 | Minor | Hard-coded threshold for BLS signature | `inference-chain/x/bls/keeper/msg_server_group_validation.go` |
| GEB-43 | Minor | Valid Dealers Can Be Less Than Threshold | `inference-chain/x/bls/keeper/phase_transitions.go` |
| GEB-38 | Minor | Inconsistent Comparison of Deadline Block | `inference-chain/x/bls/keeper/msg_server_verifier.go` |
| GEB-37 | Minor | DKG Process Can Be Stuck Due to Internal Errors | `inference-chain/x/bls/keeper/phase_transitions.go` |
| GEB-58 | Info | `ProcessThresholdSigningRequested()` Incorrectly Returns Error | `decentralized-api/internal/bls/manager.go`<br>`decentralized-api/internal/bls/threshold_signing.go` |
| GEB-50 | Info | Unused Function `computeParticipantPublicKey()` in `bls_crypto.go` | `inference-chain/x/bls/keeper/bls_crypto.go`<br>`inference-chain/x/bls/keeper/msg_server_group_validation_timing_test.go` |
| GEB-49 | Minor | Missing Check of Withdrawal and Mint Amount | `proposals/ethereum-bridge-contact/contracts/BridgeContract.sol` |
| GEB-48 | Minor | Missing Signed Status in `parseEpochDataFromJSON()` | `decentralized-api/internal/bls/verifier.go` |
| GEB-42 | Minor | Unhandled Error of `EmitTypedEvent()` | `inference-chain/x/bls/keeper/dkg_initiation.go`<br>`inference-chain/x/bls/keeper/msg_server_group_validation.go`<br>`inference-chain/x/bls/keeper/msg_server_verifier.go` |
| GEB-40 | Minor | Broken Cleanup Logic | `inference-chain/x/inference/keeper/bridge_transaction.go` |
| GEB-21 | Minor | Ineffective Polynomial Degree Check in `evaluatePolynomial()` | `decentralized-api/internal/bls/dealer.go`<br>`decentralized-api/internal/bls/dealer_test.go` |
| GEB-22 | Minor | Unchecked `amountToBytes32()` Panic on Oversized Amounts | `inference-chain/x/inference/keeper/bridge_utils.go`<br>`inference-chain/x/inference/keeper/bridge_utils_test.go` |
| GEB-08 / GEB-09 | Minor | `ADMIN` Role Cannot Not Be Updated / Migration From cw20‑base Leaves Required Wrapped‑Token State Uninitialized | `inference-chain/contracts/wrapped-token/src/contract.rs` |
| GEB-20 | Minor | Decoding Returns Zero If Fails | `inference-chain/x/inference/keeper/bridge_utils.go`<br>`inference-chain/x/inference/keeper/bridge_utils_test.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_mint.go`<br>`inference-chain/x/inference/keeper/msg_server_request_bridge_withdrawal.go` |

## Summary of fixes

- **GEB-55:** Genesis export/import now carries enough BLS state that in-flight DKG and threshold-signing work is not silently dropped across upgrades (types, keeper wiring, and tests).
- **GEB-34:** L1 bridge contract restricts who may set the genesis epoch group key so it cannot be controlled by arbitrary external callers.
- **GEB-16:** Bridge mint/withdrawal flows bind BLS-signed payloads to the intended destination (chain + contract / context) on-chain and in the reference Solidity contract, with tests.
- **GEB-14:** Threshold signing no longer relies on a user-chosen `RequestId` in a way that enables front-run poisoning; request identity is derived safely (proto + keeper + tests).
- **GEB-57:** Invalid slot ranges fail explicitly instead of being silently clamped during DKG initiation.
- **GEB-56:** `RequestThresholdSignature` validates the requested epoch against current chain state via keeper hooks (plus inference/BLS integration test updates).
- **GEB-47:** Group validation uses the proper configurable BLS threshold rather than a hard-coded fraction.
- **GEB-43:** Phase transitions ensure the count of cryptographically valid dealers meets the required threshold before proceeding.
- **GEB-38:** Verifier deadline checks use a consistent block-height comparison (inclusive vs exclusive aligned).
- **GEB-37:** Internal errors during DKG no longer leave the process wedged; transitions and error handling were corrected.
- **GEB-58:** DAPI threshold-signing request processing returns success when the operation is intentionally skipped or already handled, avoiding spurious errors upstream (`manager` + `threshold_signing`).
- **GEB-50:** Dead `computeParticipantPublicKey` code path removed / refactored with tests updated accordingly.
- **GEB-49:** Solidity bridge enforces sensible checks on mint and withdrawal amounts.
- **GEB-48:** Epoch JSON parsing records signed vs unsigned validator status so verification logic sees the right state.
- **GEB-42:** `EmitTypedEvent` errors are handled instead of ignored in DKG / group validation / verifier paths.
- **GEB-40:** Bridge transaction cleanup correctly clears related state when transactions complete or are removed.
- **GEB-21:** `evaluatePolynomial` enforces an effective polynomial degree check (with unit tests).
- **GEB-22:** `amountToBytes32` bounds inputs so oversized amounts cannot panic the keeper.
- **GEB-08 / GEB-09:** Wrapped-token contract allows admin updates where required and migration initializes required state after cw20-base migration.
- **GEB-20:** ABI decoding failures surface as errors instead of falling through to zero values; mint/withdrawal handlers use the safer decoding (with tests).

### CODE REVIEW

File: proposals/ethereum-bridge-contact/contracts/BridgeContract.sol
Comment (akup) [inline proposals/ethereum-bridge-contact/contracts/BridgeContract.sol]: Unneeded condition, it is always met

File: inference-chain/x/bls/keeper/genesis.go
Comment (akup) [inline inference-chain/x/bls/keeper/genesis.go:1]: //nolint:forbidigo

File: inference-chain/proto/inference/inference/tx.proto [hunk 1]
```diff
@@ -534,6 +534,7 @@ message MsgRequestBridgeWithdrawal {
   string user_address = 2; // User address that triggered the withdrawal
   string amount = 3; // Amount to withdraw (as string to handle large numbers)
   string destination_address = 4; // Ethereum address to receive tokens
+  string destination_bridge_address = 5; // The specific bridge contract to process the withdrawal
```
Comment (akup) [inline inference-chain/proto/inference/inference/tx.proto:384]: contract.rs at contracts/wrapped-token should be also effected. Now it will fail as contract.rs has the empty field

**Comment by 0xgonka** [comment]
a few merge conflicts now

**Comment by 0xgonka** [comment]
All 19 findings appear properly addressed. A few items to consider:

**Should fix:**
- `BridgeContract.sol` `withdraw()`: `processedRequests[cmd.epochId][cmd.requestId] = true` is set AFTER the external call (line 376). While `nonReentrant` prevents reentrancy, moving the flag before `cmd.recipient.call{value: cmd.amount}("")` follows checks-effects-interactions and is better defense-in-depth.
- `SetEffectiveEpochIndex` now requires `sdk.Context` (type assertion in epoch.go). Any caller passing a plain `context.Context` will panic at runtime. Verify all callers are covered — this is a breaking change in the function signature.

**Worth verifying:**
- GEB-55: `CurrentSigningEpochId` is only imported when `> 0` in `InitGenesis`. Epoch 0 is the default/unset state so this should be fine, but confirm epoch 0 is never a meaningful signing epoch.
- GEB-09: Migration now only allows migration from `CONTRACT_NAME` ("wrapped-token"), rejecting "crates.io:cw20-base". Confirm no existing deployed contracts still carry the cw20-base contract name — they wouldn't be migratable after this change.
- GEB-38: Deadline comparison changed from `>=` to `>`, allowing submissions AT the deadline block. Verify this is consistent with `ProcessDKGPhaseTransitionForEpoch` which uses `>=` to trigger transitions at the same block — both run in the same block but `AnteHandler`/`DeliverTx` runs before `EndBlocker`, so a submission at the deadline block would be accepted before the transition fires. This is correct but tight.

**Nits:**
- GEB-50: `computeParticipantPublicKey` is marked `// Deprecated` but still present. Consider removing entirely to avoid confusion — dead code in crypto-critical modules is a maintenance risk.
- GEB-21: The `evaluatePolynomial` fix adds explicit handling for `len(polynomial) == 1`, but the original Horner's loop was already correct for this case (loop wouldn't execute). The explicit check is cleaner but worth noting it was technically not a bug.
- `bridge_utils_test.go` — verify test coverage exists for edge cases: amounts exceeding 256 bits, negative values, malformed Ethereum addresses, and empty strings for all three conversion functions.

Overall this is a solid batch of fixes. The most important ones (GEB-34 genesis key restriction, GEB-16 destination binding, GEB-14 front-run prevention, GEB-55 genesis state preservation) are all correctly implemented with proper test coverage.

**Comment by GLiberman** [comment]
> * GEB-09: Migration now only allows migration from `CONTRACT_NAME` ("wrapped-token"), rejecting "crates.io:cw20-base". Confirm no existing deployed contracts still carry the cw20-base contract name — they wouldn't be migratable after this change.

wrapped token contract wasn't set in gov / so it wasn't used on chain yet

**Comment by 0xgonka** [comment]
this one has merge conflicts and a failure still 🙏

**Review by sam0x17 — APPROVED** [review]

**Review by 0xgonka — APPROVED** [review]
